### ISSUE TITLE
Cookies on parent domain are also set on subdomain

### ISSUE THREAD

**Issue by danielbender1989** [aws-samples/cloudfront-authorization-at-edge#183]
Hello There,

we are currently using this project to deploy a Serverless application in a dev and prod environment. We are deploying it with AWS SAM directly from the Serverless Application repository with the following configuration (we create the CloudFront distribution and the Cognito UserPool separately). CloudFormation extract we use for the deployment:

```  LambdaEdgeProtection:
    Type: AWS::Serverless::Application
    Properties:
      Location:
        ApplicationId: arn:aws:serverlessrepo:us-east-1:520945424137:applications/cloudfront-authorization-at-edge
        SemanticVersion: 2.0.13
      Parameters: 
        CreateCloudFrontDistribution: false
        UserPoolArn: !Ref UserPoolArn
        UserPoolClientId: !Ref UserPoolClient
        HttpHeaders: |-
            {
              "Content-Security-Policy": "default-src 'self'; img-src 'self'; script-src 'self' https://code.jquery.com https://stackpath.bootstrapcdn.com; style-src 'self' 'unsafe-inline' https://stackpath.bootstrapcdn.com; object-src 'self'; connect-src 'self' https://*.amazonaws.com https://*.amazoncognito.com",
              "Strict-Transport-Security": "max-age=31536000; includeSubdomains; preload",
              "Referrer-Policy": "same-origin",
              "X-XSS-Protection": "1; mode=block",
              "X-Frame-Options": "DENY",
              "X-Content-Type-Options": "nosniff"
            }
        CookieSettings: |-
            {
              "idToken": "Secure; SameSite=Lax",
              "accessToken": "Secure; SameSite=Lax",
              "refreshToken": "Secure; SameSite=Lax",
              "nonce": "Secure; HttpOnly; Max-Age=300; SameSite=Lax"
            }
```

We are deploying the code in two different environments (dev and prod) with the domain names "dev.mydomain.net" and "mydomain.net". The issue we are facing is that the Auth cookies which are set for the prod environment are inherited by the dev environment which causes in a violation of max cookie size in our Web Application Firewall.

![image](https://user-images.githubusercontent.com/39130487/152512283-064a84ad-051e-4074-9b66-236c9b2cc76d.png)

I believe this is due to the fact that the domain of the Cookies has a leading "." causing the cookies to be applied to subdomains as well which is not the behavior I would expect. We will on short term fix it by changing our dev domain to something like mydomain-dev.net. However, we would be very grateful if we can get a configuration option to store the cookies under a domain without a leading ".".

I think it should be enough to either change it in the function "withCookieDomain" or add an option to configure the domain:

```
function withCookieDomain(
  distributionDomainName: string,
  cookieSettings: string
) {
  // Add the domain to the cookiesetting
  if (cookieSettings.toLowerCase().indexOf("domain") === -1) {
    // Add leading dot for compatibility with Amplify (or js-cookie really)
    return `${cookieSettings}; Domain=.${distributionDomainName}`;
  }
  return cookieSettings;
}
``` 

Thanks in advance!
Daniel

**Comment by ottokruse** [comment]
Hi @danielbender1989 ,
Did you consider adding the domain to your cookie settings? That would solve it too I think. E.g.:
(note `Domain` added without leading `.`)
```
        CookieSettings: |-
            {
              "idToken": "Domain=mydomain.net; Secure; SameSite=Lax",
              "accessToken": "Domain=mydomain.net; Secure; SameSite=Lax",
              "refreshToken": "Domain=mydomain.net; Secure; SameSite=Lax",
              "nonce": "Domain=mydomain.net; Secure; HttpOnly; Max-Age=300; SameSite=Lax"
            }
```
So for dev and prod you would use different parameters then. You could e.g. make `CookieDomain` a CFN parameter in your template to make it nicer:

```
        CookieSettings: |-
            !Sub {
              "idToken": "Domain=${CookieDomain}; Secure; SameSite=Lax",
              "accessToken": "Domain=${CookieDomain}; Secure; SameSite=Lax",
              "refreshToken": "Domain=${CookieDomain} Secure; SameSite=Lax",
              "nonce": "Domain=${CookieDomain}; Secure; HttpOnly; Max-Age=300; SameSite=Lax"
            }
```

With regard to the default implementation if you don't provide a domain, having the leading `.` there had a reason (the code comment tells) unfortunately the code comment doesn't mention what exactly could go wrong without it. Maybe should try removing it indeed and see if we can get rid of it.

**Comment by danielbender1989** [comment]
Hi @ottokruse,

I set it up like you proposed with some little adaptions:

```
        CookieSettings: !Sub |-
            {
              "idToken": "Domain=${DomainName}; Secure; SameSite=Lax",
              "accessToken": "Domain=${DomainName}; Secure; SameSite=Lax",
              "refreshToken": "Domain=${DomainName}; Secure; SameSite=Lax",
              "nonce": "Domain=${DomainName}; Secure; HttpOnly; Max-Age=300; SameSite=Lax"
            }
```
I can also see that the settings are successfully deployed to the ParseAuthHandler and the RefreshAuthHandler in the configuration.json with the correct domain entered. However, the cookies are still saved with a leading `.`. Could it be that the function `withCookieDomain` overrides the Domain from the Cookie settings?

**Comment by ottokruse** [comment]
Mmmm the adding of the leading `.` is protected by this (crude) `if` statement, which shouldn't trigger when you explicitly include a domain in your cookie settings:

https://github.com/aws-samples/cloudfront-authorization-at-edge/blob/8b61156a37cf30999ce61fbee9fcfd344f1fb95a/src/lambda-edge/shared/shared.ts#L258-L261

I gave it a shot myself and added this config to one of my (testing) stacks:

```json
{
  "idToken":"Domain=d1kvy4j7zfgzqx.cloudfront.net; Secure; SameSite=Lax",
  "accessToken":"Domain=d1kvy4j7zfgzqx.cloudfront.net; Secure; SameSite=Lax",
  "refreshToken":"Domain=d1kvy4j7zfgzqx.cloudfront.net; Secure; SameSite=Lax",
  "nonce":"Domain=d1kvy4j7zfgzqx.cloudfront.net; Secure; HttpOnly; Max-Age=300; SameSite=Lax"
}
```

And that works flawless for me, no leading `.` in the domain of my JWT cookies anymore.

What is different in your case, can you confirm that CloudFront was redeployed with the new Lambda@Edge code?

**Comment by danielbender1989** [comment]
@ottokruse I also tested this function locally with the expected outcome - which makes me wonder as well. What I can see is the following:

1. The stack updated the lambda functions and the CloudFront distribution:
![image](https://user-images.githubusercontent.com/39130487/152539396-d9a14854-7b21-4e82-9c09-6051bceba69f.png)


2. The configuration.json in the Lambdas (e.g. ParseAuthHandler) lambda got updated:

![image](https://user-images.githubusercontent.com/39130487/152539005-713b9152-4f9b-4f20-82d0-3785c0890836.png)

The update version of the lambda function is Version 2 and I can also see that reflected in the configuration of the CloudFront distribution:

![image](https://user-images.githubusercontent.com/39130487/152539581-a241ac67-065a-45c5-8ddb-96945c7ace60.png)

I also invalidated the cache of the Cloudfront distribution:

![image](https://user-images.githubusercontent.com/39130487/152539723-0cbaca93-e074-4075-b4e4-889aa8fe4a05.png)

I have to admit that I am not super familiar with the lifecycle of Lambda@Edge functions. Could it be that although the Lambda was updated, it takes some time until the replicas get updated aswell?

Thanks a lot for your help!

**Comment by ottokruse** [comment]
> Could it be that although the Lambda was updated, it takes some time until the replicas get updated aswell?

if you run this all through CloudFormation, and update your CloudFront distribution also through CloudFormation, then when that is "UPDATE_COMPLETE" everything should be done and ready. The Lambda@Edge viewer request responses are not cached by CloudFront by the way (could be cached by your browser though -- maybe that's the snag now).

> I also tested this function locally with the expected outcome - which makes me wonder as well.

Sorry am not following, can you explain what you mean?

**Comment by danielbender1989** [comment]
> could be cached by your browser though -- maybe that's the snag now
I deleted cache + cookies and still get the same result.

> Sorry am not following, can you explain what you mean?

I wrote a little .js script to check how the function behaves - with the outcome as you describe:

```
function withCookieDomain(
  distributionDomainName,
  cookieSettings
) {
  // Add the domain to the cookiesetting
  if (cookieSettings.toLowerCase().indexOf("domain") === -1) {
    // Add leading dot for compatibility with Amplify (or js-cookie really)
    return `${cookieSettings}; Domain=.${distributionDomainName}`;
  }
  return cookieSettings;
}

var distributionDomainName = "dev.mydomain.net";
var cookieSettings = "Domain=dev.mydomain.net Secure; SameSite=Lax";

var result = "";

result = withCookieDomain(
  distributionDomainName,
  cookieSettings
)

console.log(result)
```

Which logs `Domain=dev.mydomain.net Secure; SameSite=Lax`

**Comment by ottokruse** [comment]
Reading the [MDN Set-Cookie page](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie) and the [concerned RFC](https://www.rfc-editor.org/rfc/rfc6265#section-5.3) again closely, I think it's saying that to set a host-only cookie, you need to *not* include the `Domain` attribute at all.

Which means my advice, to add a domain to the cookiesettings, is not right. But leaving it out is not gonna work either because then the if statement mentioned earlier will add it anyway.

Long story short, I now believe that if statement must be removed, it's not adding any value and only making it impossible to have host-only cookies. The entire `withCookieDomain` function can be removed then. I'll have to experiment with this some more.

**Comment by ottokruse** [comment]
Good news, this "hack" works for me, to go around the `if` statement, and get host-only cookies in FireFox and Chrome:

```
{
  "idToken":"Domain=; Secure; SameSite=Lax",
  "accessToken":"Domain=; Secure; SameSite=Lax",
  "refreshToken":"Domain=; Secure; SameSite=Lax",
  "nonce":"Domain=; Secure; HttpOnly; Max-Age=300; SameSite=Lax"
}
```

**Comment by danielbender1989** [comment]
Thanks I will try it. I also figured out now that the Cookie header works like you said but my browser actually adds the `.`.

set-cookie header of the request:

```
CognitoIdentityServiceProvider.xxx.LastAuthUser=xxx; Domain=**dev.mydomain.net**; Secure; SameSite=Lax
```

**Comment by danielbender1989** [comment]
@ottokruse this was successul! Thanks a lot for your help. To wrap it up this is our working configuration:

```
  LambdaEdgeProtection:
    Type: AWS::Serverless::Application
    Properties:
      Location:
        ApplicationId: arn:aws:serverlessrepo:us-east-1:520945424137:applications/cloudfront-authorization-at-edge
        SemanticVersion: 2.0.13
      Parameters: 
        CreateCloudFrontDistribution: false
        UserPoolArn: !Ref UserPoolArn
        UserPoolClientId: !Ref UserPoolClient
        HttpHeaders: |-
            {
              "Content-Security-Policy": "default-src 'self'; img-src 'self'; script-src 'self' https://code.jquery.com https://stackpath.bootstrapcdn.com; style-src 'self' 'unsafe-inline' https://stackpath.bootstrapcdn.com; object-src 'self'; connect-src 'self' https://*.amazonaws.com https://*.amazoncognito.com",
              "Strict-Transport-Security": "max-age=31536000; includeSubdomains; preload",
              "Referrer-Policy": "same-origin",
              "X-XSS-Protection": "1; mode=block",
              "X-Frame-Options": "DENY",
              "X-Content-Type-Options": "nosniff"
            }
        CookieSettings: |-
            {
              "idToken": "Domain=; Secure; SameSite=Lax",
              "accessToken": "Domain=; Secure; SameSite=Lax",
              "refreshToken": "Domain=; Secure; SameSite=Lax",
              "nonce": "Domain=; Secure; HttpOnly; Max-Age=300; SameSite=Lax"
            }
```
